EIPS/eip-7399.md

```diff
@@ -0,0 +1,222 @@
+
+`flashFee` reverts on unsupported assets, because returning a numerical value would be incorrect.
+
+`flashFee` returns a value that is consistent with an impossible loan when the `lender` doesn't have enough liquidity to serve the loan.
+
+`flash` has been chosen as a function name as a verb which is descriptive enough, unlikely to clash with other functions in the `lender`, and including both the use cases in which the assets lent are held or minted by the `lender`.
+
+Existing flash lenders all provide flash loans of several asset types from the same contract. Providing a `asset` parameter in both the `flash` and callback functions matches closely the observed functionality.
+
+A `bytes calldata data` parameter is included for the `initiator` to pass arbitrary information to the `receiver`. The `receiver` can pass arbitrary information back to the `initiator` using the `bytes memory` return value.
+
+A `initiator` will often be required in the callback function, which the `lender` knows as `msg.sender`. An alternative implementation which would embed the `initiator` in the `data` parameter by the caller would require an additional mechanism for the receiver to verify its accuracy, and is not advisable.
+
+A `loan receiver` is taken as a parameter to allow flexibility on the implementation of separate loan initiators, loan receivers, and callback receivers. This parameter is not passed on to the `callback receiver` on the grounds that it will be often the same as `callback receiver` and when not, it can be encoded in the `data` by the `initiator`.
+
+A `payment receiver` allows for the same flexibility on repayments as in borrows. Control flow and asset flow are independent.
+
+The `amount` will be required in the callback function, which the `lender` took as a parameter. An alternative implementation which would embed the `amount` in the `data` parameter by the caller would require an additional mechanism for the receiver to verify its accuracy, and is not advisable.
+
+A `fee` will often be calculated in the callback function, which the callback receiver must be aware of for repayment. Passing the `fee` as a parameter instead of appended to `data` is simple and effective.
+
+Arbitrary callback functions on callback receivers allows to implement different behaviours to flash loans on callback receivers without the need for encoding a function router using the `data` argument. A function call type is 24 bytes of which the first 20 bytes are the target address and the last 4 bytes are the function signature.
+
+The `amount + fee` are pushed to the `payment receiver` to allow for the segregation of asset and control flows. While a "pull" architecture is more prevalent, "push" architectures are also common. For those cases where the `lender` can't implement a "push" architecture, a simple wrapper contract can offer an ERC-7399 external interface, while using liquidity from the `lender` using a "pull" architecture.
```

Comment:
I think that push architectures are compatible with less contracts. Let's say that you have a lending pool which exposes an EIP-7399 `flash`, a `supply` and a `withdraw` entry-point. One could drain the contract by doing: flash(amount: x) then in the callback call supply(amount: x), which returns the assets to the lending pool, so the balance check passes and the flashLoan can end. Then the user is able call withdraw(amount: x) and get back all the supplied funds.

There are some ways to prevent this, but those are much less direct than implementing a pull flashLoan.